extensions/data-transfer/portability-data-transfer-facebook/src/main/java/org/datatransferproject/transfer/facebook/photos/FacebookPhotosExporter.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2018 The Data Transfer Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.datatransferproject.transfer.facebook.photos;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.restfb.types.Album;
+import com.restfb.types.Photo;
+import org.datatransferproject.spi.transfer.provider.ExportResult;
+import org.datatransferproject.spi.transfer.provider.Exporter;
+import org.datatransferproject.spi.transfer.types.ExportInformation;
+import org.datatransferproject.types.transfer.auth.AppCredentials;
+import org.datatransferproject.types.transfer.auth.TokensAndUrlAuthData;
+import org.datatransferproject.types.transfer.models.photos.PhotoAlbum;
+import org.datatransferproject.types.transfer.models.photos.PhotoModel;
+import org.datatransferproject.types.transfer.models.photos.PhotosContainerResource;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+
+public class FacebookPhotosExporter
+    implements Exporter<TokensAndUrlAuthData, PhotosContainerResource> {
+  private AppCredentials appCredentials;
+  private FacebookPhotosInterface photosInterface;
+
+  public FacebookPhotosExporter(AppCredentials appCredentials) {
+    this.appCredentials = appCredentials;
+  }
+
+  @VisibleForTesting
+  FacebookPhotosExporter(AppCredentials appCredentials, FacebookPhotosInterface photosInterface) {
+    this.appCredentials = appCredentials;
+    this.photosInterface = photosInterface;
+  }
+
+  @Override
+  public ExportResult<PhotosContainerResource> export(
+      UUID jobId, TokensAndUrlAuthData authData, Optional<ExportInformation> exportInformation) {
+    Preconditions.checkNotNull(authData);
+
+    ArrayList<PhotoAlbum> exportAlbums = new ArrayList<>();
+
+    // Get albums
+    Iterable<List<Album>> connection = getOrCreatePhotosInterface(authData).getAlbums();
+
+    for (List<Album> albums : connection) {
```

Comment:
Maybe we can add a TODO to paginate through albums so we don't have to hold them all in memory.

Reply:
Added

Reply:
import com.restfb.exception.FacebookOAuthException; 
 import org.datatransferproject.spi.transfer.types.CopyExceptionWithFailureReason; 
 import org.datatransferproject.spi.transfer.types.SessionInvalidatedException; 
 import org.datatransferproject.spi.transfer.types.UnconfirmedUserException; 
 import org.datatransferproject.spi.transfer.types.UserCheckpointedException; 
  
 public class FacebookTransferUtils { 
  
   public static FacebookOAuthException handleFacebookOAuthException(FacebookOAuthException e) 
       throws CopyExceptionWithFailureReason { 
     String message = e.getMessage(); 
     if (message != null && message.contains("the user is not a confirmed user")) { 
       throw new UnconfirmedUserException( 
           "The user account is not confirmed or deactivated", e); 
     } else if (message != null && message.contains("code 190, subcode 459")) { 
       // Throw out exception for known user checkpointed error from Graph API 
       throw new UserCheckpointedException("The user has been checkpointed", e); 
     } else if (message != null && message.contains("code 190, subcode 460")) { 
       throw new SessionInvalidatedException("The user session has been invalidated", e); 
     } else if (message != null && message.contains("code 190, subcode 463")) { 
       throw new SessionInvalidatedException("The user session has expired", e); 
     } else { 
       return e; 
     } 
   }

Reply:
> Maybe we can add a TODO to paginate through albums so we don't have to hold them all in memory.

Reply:
> Maybe we can add a TODO to paginate through albums so we don't have to hold them all in memory.

Reply:
> > Maybe we can add a TODO to paginate through albums so we don't have to hold them all in memory.
> 
>